components/match2/commons/match_group_legacy.lua

```diff
@@ -6,259 +6,368 @@
+	return {
+		['$notEmpty$'] = self.bracketType == 'team' and (prefix .. 'team') or prefix,
+		template = prefix .. 'team',
+		score = prefix .. scoreKey,
+		name = prefix,
+		displayname = prefix .. 'display',
+		flag = prefix .. 'flag',
+		win = prefix .. 'win',
+	}
+end
+
+---@param mapIndex number
+---@return table
+function MatchGroupLegacy:_getMap(mapIndex)
+	if self.mapMappings[mapIndex] then
+		return self.mapMappings[mapIndex]
 	end
 
-	_type = _args.type
-	if Logic.isEmpty(_type) then
-		error('argument \'type\' is empty')
+	local mapping = {}
+	Table.iter.forEachPair(self:getMap(), function (key, val)
+		val = val:gsub('%$1%$',mapIndex, 1)
+		mapping[key] = val
+	end)
+
+	self.mapMappings[mapIndex] = mapping
+	return mapping
+end
+
+---@return table
+function MatchGroupLegacy:getMap()
+	return {}
+end
+
+---@param isReset boolean
+---@param prefix string
+---@return table
+function MatchGroupLegacy:getDetails(isReset, prefix)
+	local detailsKey = isReset and 'resetDetails' or prefix .. 'details'
+	local details = Table.extract(self.args, detailsKey)
+	if details then
+		return (Json.parse(details))
 	end
 
-	local mapping = Legacy._getMapping(templateid)
-
-	local out = Json.stringify(mapping, {pretty = true})
-		:gsub('"([^\n:"]-)":', '%1 = ')
-		:gsub('type =', '["type"] =')
-		:gsub(' = %[(.-)%]', ' = { %1 }')
-	out = '-- Custom mapping for \'' .. templateid .. '\' from \'' .. oldTemplateid .. '\'\n'
-		.. 'local p = {}\n\n'
-		.. 'p["' .. oldTemplateid .. '"] = function() '
-		.. 'return ' .. out .. '\n'
-		.. 'end\n\n'
-		.. 'return p'
-	out = '[[Module:MatchGroup/Legacy/' .. templateid .. '|Link to mapping]]' .. '<pre class=\"selectall\">' .. out
-	return out .. '</pre>'
+	return {}
 end
 
----@param mapping table
+---@param opponentData table
 ---@return table
-function Legacy._convert(mapping)
-	local newArgs = {}
-	for source, target in pairs(mapping) do
-		-- nested tables
-		if type(target) == 'table' then
-			-- flatten nested tables like RxGx
-			local flatten = target['$flatten$'] or {}
-			local flattened = {}
-			for _, flattensource in ipairs(flatten) do
-				local toFlatten = _args[flattensource] or {}
-				if type(toFlatten) == 'string' then
-					toFlatten = Json.parse(toFlatten)
-				end
-				for key, val in pairs(toFlatten) do
-					flattened[key] = val
-				end
-			end
+function MatchGroupLegacy:readOpponent(opponentData)
+	local opponent = self:_copyAndReplace(opponentData, self.args)
+	opponent.type = self.bracketType
+	return opponent
+end
 
-			target['$flatten$'] = nil
+---@param isReset boolean
+---@param match1params match1Keys
+---@param match table
+function MatchGroupLegacy:handleOpponents(isReset, match1params, match)
+	local scoreKey = isReset and 'score2' or 'score'
+	Array.forEach(Array.range(1, MAX_NUMBER_OF_OPPONENTS), function (opponentIndex)
+		local opp = self:getOpponent(match1params['opp' .. opponentIndex], scoreKey)
+		if Logic.isEmpty(self.args[opp['$notEmpty$']]) then return end
```

Comment:
How to handle a empty `R1D1` or `R1D1team`? @hjpalpha @Rathoz

Reply:
When does that happen and what's the effect?

Reply:
Old module would fill opponentX with `{type = 'team', template = 'TBD', icon = WikiSpecific.defaultIcon, name = 'TBD'}`, this depends a bit of how match group input handle a empty opponentX

Reply:
New MGI will make `bye`-literal for empty team inputs, and blank opponent of correct type for solo/party

Reply:
Nice

Reply:
Try it out on LoL or SC2, they are using the new standard MGI opponent parsing

Reply:
Oh, and assuming it's not needed, remove any `defaultIcon` from brkts/wikispec that remains (Rocket League only?)

Reply:
Different pr :x

Reply:
LoL is now live with the new MGI

Reply:
all good

Reply:
What happens on some other wiki? eg valorant? does it break or is fine?

Reply:
not good xD

Reply:
Let's focus on rolling out the new MGI/Customs ASAP, so we can be simple here

Reply:
hmm i think if it errors now on missing opponent it did do so with the old setup too
so it shouldn't break any additional pages

nonetheless it might be good to wait for the MGI refactors be done (at least for main wikis)

Reply:
Previously it add up to 2 team opponents if non existed, new doesn't

Reply:
ag

Reply:
Ok, I wait, let this convo open

Reply:
Should go pretty fast, I'll do the implemenation of all on the list tomorrow that no1 else picked up during the evening. The mobas should only take 10 mins each

Reply:
mind there are several open prs for match2 we have to fix them too
most of those wikis have them in use despite it not being merged yet

![IMG_4108](https://github.com/user-attachments/assets/b0d6175a-71d8-43b5-9324-ea5d8bc4f409)

Reply:
i think for warcraft and stormgate sc(2) can give a good example

Reply:
they don't in the todo of this

Reply:
yeah more of a general reminder not related to this pr^^

Reply:
sc(2) is up too
so only warcraft remains untouched for now

Reply:
all merged except WC3

Reply:
Seems ok